### PR TITLE
Allowing dataclasses which require list protobuf deserialization

### PR DESCRIPTION
**Pull request by JackUrb**
## Tracking issue
closes flyteorg/flyte/issues/5593
closes flyteorg/flyte/issues/4581

## Why are the changes needed?

As described in flyteorg/flyte/issues/5593, when handling the protobuf serialization/deserialization for a dataclass, attributes that are lists end up being deserialized to lists rather than dicts (despite the function being `_MessageToDict`).

## What changes were proposed in this pull request?

The primary change is to the `ProtobufTransformer`'s `to_literal` method. If the `_MessageToDict` method returns a `list` instead of a `dict`, rather than try to construct a protobuf `Struct` (which can't be done for a list), we use a separate `_handle_list_literal` method, which ultimately treats the processing more similarly to the `ListTransformer`.

An additional change is made to the `assert_type` method to avoid an issue where the `isinstance` clause was unnecessarily evaluated (due to the logic ordering) in cases where the type `t` could be a generic (for which the assertion error `isinstance() argument 2 cannot be a parameterized generic` would be thrown). Swapping the logic order means that if `get_origin` is not None (which occurs in parameterized generics), the `isinstance` will not be called.

## How was this patch tested?

Adds a new test to the `TypeEngine` testing, `test_DataclassTransformer_for_list_attributes`.

We've also been running this change [locally](https://github.com/datologyai/flytekit/pull/1/files) for some time, and it has been stable.

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

### CODE REVIEW

File: flytekit/core/type_engine.py [hunk 1]
```diff
@@ -706,10 +706,24 @@ def tag(expected_python_type: Type[T]) -> str:
...
     def to_literal(self, ctx: FlyteContext, python_val: T, python_type: Type[T], expected: LiteralType) -> Literal:
         struct = Struct()
         try:
-            struct.update(_MessageToDict(cast(Message, python_val)))
+            message_dict = _MessageToDict(cast(Message, python_val))
+            try:
+                struct.update(message_dict)
+            except Exception:
+                if isinstance(message_dict, list):
+                    return self._handle_list_literal(ctx, message_dict)
```
Comment (pingsutw) [inline flytekit/core/type_engine.py:725]: nit: could we do this?
```python
if isinstance(message_dict, list):
    return self._handle_list_literal(ctx, message_dict)
else:
    struct.update(message_dict)
```

**Review by pingsutw — COMMENTED** [review]
thanks. overall lgtm. Some tests are failing in python 3.9. mind taking a look

File: flytekit/core/type_engine.py (hunk 1 above)
Comment (JackUrb) [reply flytekit/core/type_engine.py:725]: Seems like yes, will update

**Comment by JackUrb** [comment]
@pingsutw I'm having a hard time telling quite why the test is failing in the narrow case of py3.9 tensorflow file conversion. Is there any retry logic somewhere else in flyte that effectively checks for successful conversion, and then does something else on failure? It's hard to navigate why this conversion is this specific case breaks, and the `FlyteFile` `_SpecificFormatClass` portion is pretty hard to understand.

**Comment by JackUrb** [comment]
Note, it's possible this is entirely superseded by #2600 - happy to assist on that effort if extra hands would help.

**Comment by Future-Outlier** [comment]
> Note, it's possible this is entirely superseded by #2600 - @Future-Outlier happy to assist on that effort if extra hands would help.

@JackUrb 
I just back from the military service!
I'm in charge of JSON IDL, do you want to collaborate with me? or help review my PR?
I can hop on a call with you.

**Comment by JackUrb** [comment]
Indeed, this entire issue is resolved by #5742. Very excited for v1.14.0 to come out! Thanks @Future-Outlier

**Pull request closed without merging**